src/Kvasir/Reconstitution/Lookups.cs

```diff
@@ -30,19 +30,30 @@ internal static partial class Lookup {
         public static object ByKey(DBData key, IEnumerable<object> entities, DataExtractionPlan extractor) {
             Debug.Assert(entities.All(e => e.GetType().IsInstanceOf(extractor.ExpectedSource)));
 
+            if (CacheDictionary.ContainsKey(key))
+                return CacheDictionary[key];
+
             foreach (var entity in entities) {
+                if (CacheValues.Contains(entity))
```

Comment:
I think we might be able to do
```
foreach (var entity in entities.Where(e => !cacheValues_.Contains(e)))
```
to avoid having to do a check + `continue` in the body of the loop. (Not 100% sure what the performance implications of pulling in LINQ are, but I like the fluent API it provides and try to use it where possible. Especially if it makes things cleaner.)

Reply:
A minor counterargument is that this suggestion checks _every_ entity against the cache, whereas my original only performs the checks on the entities with index lesser than the result. An alternative would be to wrap everything inside of the `if` statement, which is functionally equivalent but is one line shorter. I'll put such an alternative in my updates.

Reply:
It actually doesn't - LINQ is lazily evaluated. The `Where` extension method is essentially equivalent to this:
```
IEnumerable<T> Where<T>(this IEnumerable<T> self, Predicate<T> predicate) {
    foreach (var element in self) {
        if (predicate(element)) {
            yield element;
        }
    }
}
```

This creates an "anonymous IEnumerable" where the `yield` statement produces a coroutine that "yield" control back to the caller until the next time the pseudo-Enumerable's iterator is advanced, at which point it picks up back at the most recent yield point. So in both situations, we only evaluate the predicate (in this case, the `.Contains(...)` check) on items in the collection up to and including the eventual result.

Reply:
That's neat! This works well then.